### PR TITLE
Sd utilize

### PR DESCRIPTION
**Pull request by batphonghan**
### What's in this PR:

* SD utilization integration

### CODE REVIEW

**Review by rahuljaguste — DISMISSED** [review]

File: shared/types/api/node.go [hunk 1]
```diff
@@ -91,17 +92,17 @@ type NodeDepositSdAllowanceResponse struct {
 }
 
 type CanNodeDepositResponse struct {
-	Status                   string         `json:"status"`
-	Error                    string         `json:"error"`
-	CanDeposit               bool           `json:"CanDeposit"`
-	InsufficientBalance      bool           `json:"insufficientBalance"`
-	InvalidAmount            bool           `json:"invalidAmount"`
-	DepositPaused            bool           `json:"depositPaused"`
-	NotEnoughSdCollateral    bool           `json:"notEnoughSdCollateral"`
```
[the remark this answers is not in the document]
Comment (batphonghan) [reply shared/types/api/node.go:100]: It's not remove bro, `NotEnoughSdCollateral` now inside `SdStatusResponse`

File: stader-cli/validator/deposit.go [hunk 2]
```diff
@@ -52,22 +56,117 @@ func nodeDeposit(c *cli.Context) error {
...
+	}
+
+	fmt.Printf(
+		"The node %s%s%s has a balance of %.6f SD.\n\n",
+		log.ColorBlue,
+		status.AccountAddress,
+		log.ColorReset,
+		math.RoundDown(eth.WeiToEth(status.AccountBalances.Sd), Decimal))
+
+	canNodeDepositResponse, err := staderClient.CanNodeDeposit(baseAmount, big.NewInt(0), big.NewInt(int64(numValidators)), true)
```
[the remark this answers is not in the document]
Comment (batphonghan) [reply stader-cli/validator/deposit.go:72]: Thank you bro, `canDeposit ` is missing `utilityAmount `. I will update it.

File: stader-cli/node/repay-exess-sd.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	_, err = staderClient.CanNodeRepaySd(amountWei)
+	if err != nil {
+		return err
+	}
+
+	res, err := staderClient.NodeRepaySd(amountWei)
+	if err != nil {
+		return err
+	}
```
Comment (bharath-123) [inline stader-cli/node/repay-exess-sd.go:84]: This should be `canNodeRepayExcess` and `NodeRepayExcessSd`. Also let's remove the `Node` prefix. it seems unecesssary

File: stader/api/validator/deposit.go [hunk 4]
```diff
@@ -2,20 +2,77 @@ package validator
...
 	_ "golang.org/x/sync/errgroup"
-	"math/big"
 
 	"github.com/stader-labs/stader-node/shared/services"
 	"github.com/stader-labs/stader-node/shared/types/api"
 	"github.com/stader-labs/stader-node/shared/utils/eth1"
 	"github.com/stader-labs/stader-node/shared/utils/validator"
 )
 
+const MaxBondThreshold = 2
```
Comment (bharath-123) [inline stader/api/validator/deposit.go:23]: is this 1Eth worth SD here?

File: stader/api/node/sd-status.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	totalValidatorKeys, err := node.GetTotalValidatorKeys(prn, operatorID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	totalValidatorNonTerminalKeys, err := node.GetTotalNonTerminalValidatorKeys(prn, nodeAccount.Address, totalValidatorKeys, nil)
```
Comment (bharath-123) [inline stader/api/node/sd-status.go:57]: If user is running `deposit` should be consider `totalValidatorNonTerminalKeys + validatorsToCreate` ?

Comment (bharath-123) [reply stader/api/node/sd-status.go:57]: in that case, we will have to pass in the amount of validators we are creating. If we are not creating, then it should be 0

Comment (bharath-123) [reply stader/api/node/sd-status.go:57]: cool i think its considering the validators post addition below.

File: stader-cli/node/send-el-rewards.go [hunk 6]
```diff
@@ -2,6 +2,7 @@ package node
 
 import (
 	"fmt"
+
```
Comment (bharath-123) [inline stader-cli/node/send-el-rewards.go:5]: let's see if we can avoid getting these changes to the PR. creates unnecessary diff. 

I think it might be part of a lint tho

File: shared/services/state/network-state.go [hunk 7]
```diff
@@ -497,6 +516,7 @@ func CreateMetricsCache(
 	metricsDetails.UnclaimedSocializingPoolElRewards = math.RoundDown(eth.WeiToEth(rewardClaimData.unclaimedEth), SixDecimalRound)
 	metricsDetails.UnclaimedSocializingPoolSDRewards = math.RoundDown(eth.WeiToEth(rewardClaimData.unclaimedSd), SixDecimalRound)
 
+	metricsDetails.OperatorSDUtilized = math.RoundDown(eth.WeiToEth(sdUtilized), SixDecimalRound)
```
Comment (bharath-123) [inline shared/services/state/network-state.go:519]: we might also want to show the interest that the user has accrued. I think the interested will be accrued on top of the users Utilizer balance

File: shared/types/api/node.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bharath-123) [reply shared/types/api/node.go:100]: cool bro. makes sense

File: shared/types/api/node.go [hunk 8]
```diff
@@ -265,11 +265,12 @@ type SendElRewardsResponse struct {
...
-	Status                     string         `json:"status"`
-	Error                      string         `json:"error"`
-	InsufficientSdCollateral   bool           `json:"insufficientSdCollateral"`
-	InsufficientWithdrawableSd bool           `json:"insufficientWithdrawableSd"`
-	GasInfo                    stader.GasInfo `json:"gasInfo"`
+	Status                     string            `json:"status"`
+	Error                      string            `json:"error"`
+	InsufficientSdCollateral   bool              `json:"insufficientSdCollateral"`
+	InsufficientWithdrawableSd bool              `json:"insufficientWithdrawableSd"`
+	SdStatusResponse           *SdStatusResponse `json:"sdStatusResponse"`
```
Comment (bharath-123) [inline shared/types/api/node.go:272]: lets try to seperate this into a seperate get-sd-status cmd. It looks much cleaner like that according to me.

File: stader-cli/node/commands.go [hunk 9]
```diff
@@ -368,6 +367,67 @@ func RegisterCommands(app *cli.App, name string, aliases []string) {
 					return nodeApproveSd(c)
 				},
 			},
+			{
+				Name:      "utilize-sd",
+				Aliases:   []string{"us"},
+				Usage:     "Utilize from the Utility pool.",
+				UsageText: "stader-cli node utilization-sd",
```
Comment (bharath-123) [inline stader-cli/node/commands.go:374]:
```suggestion
				UsageText: "stader-cli node utilize-sd",
```

File: stader-cli/node/commands.go [hunk 10]
```diff
@@ -368,6 +367,67 @@ func RegisterCommands(app *cli.App, name string, aliases []string) {
...
+			{
+				Name:      "repay-excess-sd",
+				Aliases:   []string{"res"},
+				Usage:     "Repay utilized SD using the excess SD collateral",
+				UsageText: "stader-cli node repay-excess-sd [options]",
+				Flags: []cli.Flag{
+					cli.StringFlag{
+						Name:  "amount, a",
+						Usage: "The amount of SD to repay excess",
+					},
```
Comment (bharath-123) [inline stader-cli/node/commands.go:416]: No need to pass in the amount in `repay-excess-sd` . it will repay the sd collateral which is above 200% back to the pool

File: stader-cli/node/repay-exess-sd.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+	// Print what network we're on
+	err = cliutils.PrintNetwork(staderClient)
+	if err != nil {
+		return err
+	}
+
+	amountInString := c.String("amount")
```
Comment (bharath-123) [inline stader-cli/node/repay-exess-sd.go:35]: we should compute this amount. This should be the amount of SD above the 200% mark. We should repay that amount

File: stader-cli/validator/deposit.go [hunk 12]
```diff
@@ -52,22 +54,82 @@ func nodeDeposit(c *cli.Context) error {
...
+				cliutils.PrintError("You do not had enough SD in your wallet. Please deposit and try again")
+				return nil
+			}
+
+			err = node.DepositSdWithAmount(staderClient, amountToCollateral, autoConfirm, 0)
+			if err != nil {
+				return err
+			}
+
+			return nodeDeposit(c)
```
Comment (bharath-123) [inline stader-cli/validator/deposit.go:104]: i think we can just let the flow continue after this rather then re-calling the function.

File: stader/api/validator/deposit.go [hunk 13]
```diff
@@ -193,7 +246,14 @@ func canNodeDeposit(c *cli.Context, amountWei *big.Int, numValidators *big.Int,
 		return nil, fmt.Errorf("error checking for nonce override: %w", err)
 	}
 
-	gasInfo, err := node.EstimateAddValidatorKeys(prn, pubKeys, preDepositSignatures, depositSignatures, opts)
+	gasInfo, err := node.EstimateAddValidatorKeys(
+		prn,
+		big.NewInt(0),
```
Comment (bharath-123) [inline stader/api/validator/deposit.go:251]: this should be `amountUtilityWei`

File: stader/api/validator/deposit.go [hunk 14]
```diff
@@ -2,21 +2,86 @@ package validator
...
+		SdBalance:                 sdBalance,
+		SdCollateralCurrentAmount: sdCollateralCurrentAmount,
+		SdCollateralRequireAmount: minimumSDToBond,
+		SdMaxUtilizableAmount:     sdMaxCollateralAmount,
+		SdMaxCollateralAmount:     maxSDToUtilize,
+		PoolAvailableSDBalance:    poolAvailableSDBalance,
+	}, nil
+}
+
+func canNodeDeposit(c *cli.Context, amountBaseWei, amountUtilityWei, numValidators *big.Int, reloadKeys bool) (*api.CanNodeDepositResponse, error) {
```
Comment (bharath-123) [inline stader/api/validator/deposit.go:84]: make it `utilityAmountWei` to keep `canNodeDeposit` similar to `nodeDeposit`

**Review by rahuljaguste — APPROVED** [review]

File: shared/services/ec-manager.go [hunk 15]
```diff
@@ -541,12 +541,12 @@ func (p *ExecutionClientManager) Version() (string, error) {
 		Jsonrpc string   `json:"jsonrpc"`
 		Method  string   `json:"method"`
 		Params  []string `json:"params"`
-		Id      int64    `json:"id"`
+		ID      int64    `json:"id"`
```
[the remark this answers is not in the document]
Comment (batphonghan) [reply shared/services/ec-manager.go:544]: Just for fix the warning from golangci-lint
